libraries/AC_PrecLand/AC_PrecLand.cpp

```diff
@@ -692,13 +696,58 @@ bool AC_PrecLand::construct_pos_meas_using_rangefinder(float rangefinder_alt_m,
     return false;
 }
 
+// Set the active inertial frame, based on either sensor timestamp (higher priority) or PLND_LAG parameter
+// Return true if sensor does not send timestamp and front of buffer is used, or if sensor timestamp is successfully synced
+// Return false otherwise, so invalid sync is not propagated
+bool AC_PrecLand::sync_frames(const uint32_t los_meas_time_ms)
+{
+    uint64_t minimum_delta_us = UINT32_MAX * 1000UL;
+    uint16_t nearest_index = 0;  // If timestamp isn't set, use the oldest inertial buffer frame
+    // Retrieve timestamp from backend measurement.  If not zero, attempt to correlate with inertial history.
+    const uint64_t los_meas_time_us = los_meas_time_ms * 1000UL;  // overflows after approximately 49.7 days
+    if (los_meas_time_us != 0UL) {
+        // Search through inertial history for the nearest frame to the measurement timestamp
+        for (uint8_t i=0; i<_inertial_history->available(); i++) {
+            uint64_t delta_us;
+            uint64_t ih_point_us = (*_inertial_history)[i]->time_usec; // overflows after approximately 584 554.531 years
+            while (ih_point_us >= (UINT32_MAX + 1UL) * 1000UL) {
```

Comment:
A comment explaining that this is needed because of the comparison with "ms" time stamp given by the backend would be helpful

Reply:
Why does the timestamp of the companion Computer Precision landing Vision messages has a half a second delay when compared with the IMU timestamps? It causes problems when comparing the timestamps in the precision landing companion autotest.
I would like to fix that, but I need help. Does this have to do with TIME_SYNC messages? Is the delay in there purposeful?

Reply:
Yes, this is a followup of #26238. I moved the NFCish things into that one.

I added the comment that you asked.

Reply:
The wrap code has been tested with this:
```python
#!/bin/env python

UINT32_MAX = 2**32 - 1

def compare_timestamps(los_meas_time_ms, ih_point_us):
    """
    Compare two timestamps and calculate the difference in microseconds,
    accounting for 32-bit wrap-around.

    Args:
        los_meas_time_ms (int): The first timestamp in milliseconds.
        ih_point_us (int): The second timestamp in microseconds.

    Returns:
        int: The difference between the two timestamps in microseconds.

    Note:
        This function assumes that the timestamps are close to each other and
        handles the case where the difference exceeds the maximum value of a
        32-bit unsigned integer.
    """
    # simulate 32bit wrap in python
    while los_meas_time_ms >= (UINT32_MAX + 1):
        los_meas_time_ms -= (UINT32_MAX + 1)
    los_meas_time_us = los_meas_time_ms * 1000

    while ih_point_us >= (UINT32_MAX + 1) * 1000:
        ih_point_us -= (UINT32_MAX + 1) * 1000

    if ih_point_us > los_meas_time_us:
        delta_us = ih_point_us - los_meas_time_us
    else:
        delta_us = los_meas_time_us - ih_point_us
    if delta_us > UINT32_MAX * 1000 / 2:
        delta_us = (UINT32_MAX + 1) * 1000 - delta_us
    print(delta_us)
    return delta_us

assert compare_timestamps(0, 0) == 0
assert compare_timestamps(1, 0) == 1000
assert compare_timestamps(0, 1) == 1
assert compare_timestamps(2, 0) == 2000
assert compare_timestamps(0, 2) == 2
assert compare_timestamps(2, 0) == 2000
assert compare_timestamps(UINT32_MAX, UINT32_MAX*1000) == 0
assert compare_timestamps(UINT32_MAX, UINT32_MAX*1000+1) == 1
assert compare_timestamps((UINT32_MAX+1)+0, (UINT32_MAX+1)*1000+0) == 0
assert compare_timestamps((UINT32_MAX+1)+0, (UINT32_MAX+1)*1000-1) == 1
assert compare_timestamps((UINT32_MAX+1)+0, (UINT32_MAX+1)*1000-2) == 2
assert compare_timestamps((UINT32_MAX+1)+0, (UINT32_MAX+1)*1000+1) == 1
assert compare_timestamps((UINT32_MAX+1)+0, (UINT32_MAX+1)*1000+2) == 2
assert compare_timestamps((UINT32_MAX+1)+0, 2*(UINT32_MAX+1)*1000+2) == 2
assert compare_timestamps((UINT32_MAX+1)+0, 3*(UINT32_MAX+1)*1000+4) == 4
assert compare_timestamps((UINT32_MAX+0)+0, 3*(UINT32_MAX+1)*1000+4) == 1004
assert compare_timestamps((UINT32_MAX+1)+7, 3*(UINT32_MAX+1)*1000+4) == 6996
```